api/v1_projects.inc

```diff
@@ -271,6 +272,94 @@ function render_project_page_json($row)
     ];
 }
 
+// POST v1/projects/:projectid/rounds/:roundid
+// checkout a new page to proofread
+function api_v1_project_round($method, $data, $query_params)
+{
+    global $pguser;
+
+    $project = $data[":projectid"];
+    $round = $data[":roundid"];
+    check_project_user($project, $round->id);
+
+    try {
+        $lpage = get_available_page($project->projectid, $project->state, $pguser, $err);
+    } catch (Exception $exception) {
+        throw new BadRequest($exception->getMessage());
+    }
+
+    if (is_null($lpage)) {
+        throw new BadRequest($err);
+    }
+    return ["pagename" => $lpage->imagefile, "pagestate" => $lpage->page_state];
+}
+
+// PUT projects/:projectid/pages/:pagename/rounds/:roundid
+// the PUTs all can change the page text or page state.
+// they are idempotent in the sense that if the identical request is issued
+// again, either the the state and text will not be changed or the request will
+// be rejected because the page state has changed so there will be no change
+// to the server. Logging side-effects excluded.
+function api_v1_project_page_round($method, $data, $query_params)
+{
+    global $pguser;
+
+    $project = $data[":projectid"];
+    $round = $data[":roundid"];
+    check_project_user($project, $round->id);
+
+    try {
+        $lpage = get_indicated_LPage($project->projectid, $project->state, $data[":pagename"], $query_params["pagestate"], 0);
+
+        // can throw InvalidArgumentException
+        $action = get_enumerated_param($query_params, "action", null, ["return", "save_temp", "save_done", "open", "revert"]);
+
+        switch ($action) {
+        case "return":
+            $lpage->returnToRound($pguser);
+            return [];
+            break;
+        case "save_temp":
+            $lpage->saveAsInProgress(api_get_request_body(), $pguser);
+            return render_page_data($lpage);
+            break;
+        case "save_done":
+            [$page_saved, $limit_reached] = $lpage->attemptSaveAsDone(api_get_request_body(), $pguser);
+            return ["pageSaved" => $page_saved, "limitReached" => $limit_reached];
+        case "open":
+            $lpage->resume_saved_page($pguser);
+            return render_page_data($lpage);
+            break;
+        case "revert":
+            $lpage->saveAsInProgress(api_get_request_body(), $pguser);
+            $lpage->revertToOriginal();
+            return render_page_data($lpage);
+            break;
+        }
+    } catch (Exception $exception) {
+        throw new BadRequest($exception->getMessage());
+    }
+}
+
+function render_page_data($lpage)
+{
+    global $projects_url;
+    return [
+        "text" => $lpage->get_text(),
+        "imageUrl" => "$projects_url/{$lpage->projectid}/{$lpage->imagefile}",
+        "pagestate" => $lpage->page_state,
+        "canRestore" => $lpage->can_be_reverted_to_last_save(),
+    ];
+}
+
+function check_project_user($project, $round_id)
```

Comment:
`validate_user_can_work_in_project()` or something similar would be a much clearer name. I also suggest that this take a `Round` object instead of a `$round_id` since it takes a `Project` object already.

Reply:
changed name but now we don't always have a round object ready made leaving it as roundid